src/msg/async/frames_v2.h

```diff
@@ -338,25 +382,23 @@ struct SignedEncryptedFrame : public PayloadFrame<T, Args...> {
       session_stream_handlers.tx->authenticated_encrypt_update(
           std::move(c.payload));
       c.payload = session_stream_handlers.tx->authenticated_encrypt_final();
+      // append zeroized epilogue flags. CLEANME.
+      c.payload.append_zero(1);
+    } else {
+      epilogue_block_t epilogue;
+      ::memset(&epilogue, 0, sizeof(epilogue));
+
+      ceph::bufferlist::const_iterator hdriter(&c.payload, FRAME_PREAMBLE_SIZE);
+      epilogue.integrity.crc[SegmentIndex::Frame::PAYLOAD] =
+          hdriter.crc32c(hdriter.get_remaining(), -1);
+      c.payload.append(reinterpret_cast<const char*>(&epilogue), sizeof(epilogue));
     }
     return c;
   }
 
   static T Decode(ceph::crypto::onwire::rxtx_t &session_stream_handlers,
                   ceph::bufferlist &payload) {
-    if (!session_stream_handlers.rx) {
```

Comment:
Why did you remove this code block (13 lines)? Where does the decryption path lives now?

Reply:
Decryption is performed by `::handle_read_frame_segment()`. I'm working on separation of *stream crypto processing* from *frame serialization*. They appear conceptually distant. The junction point I see is the epilogue crafting (+ maybe the debug information whether a particular frame is encryptable or not). Maybe `Frame::get_buffer(...)` taking a *processor* and calling and polymorphic `::fill_epilogue(buffers)` could do.